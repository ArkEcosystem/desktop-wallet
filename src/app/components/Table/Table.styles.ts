export const defaultTableStyle = `

	table {
		margin: 0;
		padding: 0;
		width: 100%;
		border: none;

		th {
			margin: 0;
			padding-bottom: 12px;
		}

        tr:last-child, tr:last-child > td {
            border-bottom: none;
        }
	}
`;
