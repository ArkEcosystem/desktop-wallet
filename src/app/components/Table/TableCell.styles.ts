import tw from "twin.macro";

const baseStyle = tw`flex items-center px-3 my-1 transition-colors duration-100`;

const getHeight = () => tw`h-16`;

const getVariant = (variant: "start" | "middle" | "end"): any => {
	if (variant === "start") {
		return tw`pl-4 -ml-4 rounded-l-lg`;
	}

	if (variant === "end") {
		return tw`pr-4 -mr-4 rounded-r-lg`;
	}
};

export const getStyles = ({ variant }: any) => [baseStyle, getHeight(), getVariant(variant)];
