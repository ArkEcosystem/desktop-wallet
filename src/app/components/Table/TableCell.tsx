import React from "react";
import { styled } from "twin.macro";

import { getStyles } from "./TableCell.styles";

type TableCellProps = {
	variant?: "start" | "middle" | "end";
	className?: string;
	innerClassName?: string;
	isCompact?: boolean;
	children: React.ReactNode;
};

const TableCellInnerWrapper = styled.div<TableCellProps>(getStyles);

export const TableCell = ({ variant, className, innerClassName, isCompact, children }: TableCellProps) => (
	<td className={className}>
		<TableCellInnerWrapper variant={variant} className={innerClassName} isCompact={isCompact}>
			{children}
		</TableCellInnerWrapper>
	</td>
);

TableCell.defaultProps = {
	variant: "middle",
};
