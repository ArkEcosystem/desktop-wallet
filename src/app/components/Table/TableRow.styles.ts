import tw, { css } from "twin.macro";

import { TableRowFunction } from "./TableRow";

const baseStyle = tw`transition-colors duration-100`;

const getCursorStyles = (onClick?: TableRowFunction): any => (onClick ? tw`cursor-pointer` : "");

const getBorderStyles = (border?: boolean): any =>
	border ? tw`border-b last:border-b-0 border-dashed border-theme-secondary-300 dark:border-theme-secondary-800` : "";

const getHoverStyles = (isSelected?: boolean): any =>
	css`
		&:hover td > div {
			${isSelected
				? tw`bg-theme-success-100 dark:bg-theme-success-900`
				: tw`bg-theme-secondary-100 dark:bg-black`}
		}
	`;

export const getStyles = ({ onClick, border, isSelected }: any) => {
	const styles = [baseStyle, getBorderStyles(border), getCursorStyles(onClick)];

	if (onClick) {
		styles.push(getHoverStyles(isSelected));
	}

	return styles;
};
