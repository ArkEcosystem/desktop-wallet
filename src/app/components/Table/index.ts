export * from "./Table";
export * from "./TableCell";
