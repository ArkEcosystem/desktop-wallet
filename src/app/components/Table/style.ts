import { styled } from "twin.macro";

export const TableWrapper = styled.div`
	padding: 1rem;

	table {
		margin: 0;
		padding: 0;
		width: 100%;
		border: none;

		th {
			margin: 0;
		}
		td {
			vertical-align: middle;
		}

		tbody:before {
			content: "-";
			display: block;
			line-height: 1.4em;
			color: transparent;
		}
	}
`;
