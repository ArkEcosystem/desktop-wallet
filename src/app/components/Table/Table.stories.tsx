import { Address } from "app/components/Address";
import { Avatar } from "app/components/Avatar";
import { Circle } from "app/components/Circle";
import { Icon } from "app/components/Icon";
import { Table } from "app/components/Table";
import React from "react";

export default {
	title: "App / Components / Table",
};

export const Default = () => {
	const data = [
		{
			coin: "Ark",
			avatarId: "test",
			walletName: "My wallet",
			address: "ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT",
			balance: "2,000 ARK",
			fiat: "43,086.08 USD",
		},
		{
			coin: "Btc",
			avatarId: "test",
			address: "ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT",
			balance: "100 BTC",
			fiat: "1000000 USD",
		},
	];

	const columns = [
		{
			Header: "Asset Type",
			accessor: "avatarId",
		},
		{
			Header: "Wallet Address",
			accessor: "address",
		},
		{
			Header: "Balance",
			accessor: "balance",
			className: "float-right",
		},
		{
			Header: "Fiat Value",
			accessor: "fiat",
			className: "float-right",
		},
	];

	return (
		<div>
			<div>
				<Table columns={columns} data={data}>
					{(rowData: any) => (
						<tr className="border-b border-theme-neutral-200">
							<td className="py-4 mt-1">
								<div className="flex">
									<Circle>
										<Icon name={rowData.coin} />
									</Circle>
									<Avatar address={rowData.address} />
								</div>
							</td>
							<td className="py-1">
								<Address walletName={rowData.walletName} address={rowData.address} />
							</td>
							<td className="py-1 text-right text-bold">
								<div>{rowData.balance}</div>
							</td>
							<td className="py-1 text-right text-bold text-theme-neutral-400">
								<div>{rowData.fiat}</div>
							</td>
						</tr>
					)}
				</Table>
			</div>
		</div>
	);
};
