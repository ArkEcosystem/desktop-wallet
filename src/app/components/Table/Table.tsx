import React, { useMemo } from "react";
import { useSortBy, useTable } from "react-table";
import { styled } from "twin.macro";

import { Icon } from "../Icon";
import { defaultTableStyle } from "./Table.styles";

type TableProps = {
	children?: any;
	className?: string;
	data: any[];
	columns: any[];
};

const TableWrapper = styled.div`
	${defaultTableStyle}
`;

export const Table = ({ children, data, columns }: TableProps) => {
	const tableData = useMemo(() => data, [data]);
	const tableColumns = useMemo(() => columns, [columns]);

	const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } = useTable(
		{
			data: tableData,
			columns: tableColumns,
		},
		useSortBy,
	);

	const renderChildNode = (data: any, index: number) => {
		if (typeof children === "function") return children(data, index);
		return <tr />;
	};

	const getSortIconName = (isSorted: boolean, isSortedDesc: boolean) => {
		if (isSorted && isSortedDesc) {
			return "ArrowDown";
		}

		if (isSorted && !isSortedDesc) {
			return "ArrowUp";
		}

		return "Sort";
	};

	return (
		<TableWrapper {...getTableProps()}>
			<table className="table-auto">
				<thead>
					{headerGroups.map((headerGroup: any, index: number) => (
						<tr key={index} {...headerGroup.getHeaderGroupProps()}>
							{headerGroup.headers.map((column: any, thIndex: number) => (
								<th
									key={thIndex}
									className="text-sm text-left select-none text-theme-neutral-500"
									data-testid={`table__th--${thIndex}`}
									{...column.getHeaderProps(column.getSortByToggleProps())}
								>
									<div className={`flex flex-inline align-top ${column.className}`}>
										<div className="mr-2">{column.render("Header")}</div>
										<div
											className="flex items-center text-theme-color-neutral-500"
											data-testid={`table__${getSortIconName(
												column.isSorted,
												column.isSortedDesc,
											)}`}
										>
											{column.canSort && (
												<Icon
													name={getSortIconName(column.isSorted, column.isSortedDesc)}
													width={column.isSorted ? 10 : 15}
													height={column.isSorted ? 10 : 12}
												/>
											)}
										</div>
									</div>
								</th>
							))}
						</tr>
					))}
				</thead>

				<tbody {...getTableBodyProps()}>
					{rows.map((row: any) => {
						prepareRow(row);
						return { ...renderChildNode(row.original, row.index), ...row.getRowProps() };
					})}
				</tbody>
			</table>
		</TableWrapper>
	);
};

Table.defaultProps = {
	data: [],
	columns: [],
};
