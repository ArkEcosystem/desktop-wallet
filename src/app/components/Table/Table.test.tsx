import React from "react";
import { act } from "react-dom/test-utils";
import { fireEvent, render, within } from "testing-library";

import { Table } from "./Table";

const data = [
	{
		col1: "column 1",
		col2: "column 2",
		col3: "column 3",
	},
];

const columns = [
	{
		Header: "Header 1",
		accessor: "col1",
		minimumWidth: true,
		className: "flex-row-reverse",
	},
	{
		Header: "Header 2",
		accessor: "col2",
		className: "no-border",
	},
	{
		Header: "Header 3",
		accessor: "col3",
		className: "justify-end",
	},
];

describe("Table", () => {
	it("should render", () => {
		const { container } = render(
			<Table columns={columns} data={data}>
				{() => (
					<tr>
						<td>1</td>
						<td>2</td>
					</tr>
				)}
			</Table>,
		);
		expect(container).toMatchSnapshot();
	});

	it("should render empty rows if template not provided", () => {
		const { container } = render(<Table columns={columns} data={data} />);
		expect(container).toMatchSnapshot();
	});

	it("should change sort order on th click", () => {
		const { getAllByRole, getByRole } = render(<Table columns={columns} data={data} />);

		const th = getAllByRole("columnheader")[0];

		act(() => {
			fireEvent.click(th);
		});

		expect(th).toHaveTextContent("chevron-down.svg");

		expect(within(th).getByRole("img")).toHaveClass("rotate-180");

		act(() => {
			fireEvent.click(th);
		});

		expect(within(th).getByRole("img")).not.toHaveClass("rotate-180");
	});

	it("should hide header", () => {
		const { queryAllByRole } = render(<Table hideHeader columns={columns} data={data} />);
		expect(queryAllByRole("columnheader")).toHaveLength(0);
	});

	it("should render with width class applied to column header", () => {
		const { getByTestId } = render(
			<Table
				columns={[
					{
						Header: "Header 1",
						cellWidth: "width",
					},
				]}
				data={data}
			/>,
		);

		expect(getByTestId("table__th--0")).toHaveClass("width");
	});
});
