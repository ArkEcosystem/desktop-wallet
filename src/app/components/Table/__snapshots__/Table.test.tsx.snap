// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render 1`] = `
<div>
  <div
    class="sc-AxirZ jAePJ"
    role="table"
  >
    <table
      class="table-auto"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            class="text-theme-neutral-400 text-left text-xs select-none pb-3}"
            colspan="1"
            data-testid="table__th--0"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="flex flex-inline align-top  undefined"
            >
              <div
                class="mr-1"
              >
                Header 1
              </div>
              <div
                class="mt-1 text-theme-color-neutral-500"
                data-testid="table__Sort"
              >
                <div
                  class="sc-AxjAm iepfoc"
                  height="10"
                  width="10"
                >
                  <svg>
                    sort.svg
                  </svg>
                </div>
              </div>
            </div>
          </th>
          <th
            class="text-theme-neutral-400 text-left text-xs select-none pb-3}"
            colspan="1"
            data-testid="table__th--1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="flex flex-inline align-top  undefined"
            >
              <div
                class="mr-1"
              >
                Header 2
              </div>
              <div
                class="mt-1 text-theme-color-neutral-500"
                data-testid="table__Sort"
              >
                <div
                  class="sc-AxjAm iepfoc"
                  height="10"
                  width="10"
                >
                  <svg>
                    sort.svg
                  </svg>
                </div>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr>
          <td>
            1
          </td>
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render empty rows if template not provided 1`] = `
<div>
  <div
    class="sc-AxirZ jAePJ"
    role="table"
  >
    <table
      class="table-auto"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            class="text-theme-neutral-400 text-left text-xs select-none pb-3}"
            colspan="1"
            data-testid="table__th--0"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="flex flex-inline align-top  undefined"
            >
              <div
                class="mr-1"
              >
                Header 1
              </div>
              <div
                class="mt-1 text-theme-color-neutral-500"
                data-testid="table__Sort"
              >
                <div
                  class="sc-AxjAm iepfoc"
                  height="10"
                  width="10"
                >
                  <svg>
                    sort.svg
                  </svg>
                </div>
              </div>
            </div>
          </th>
          <th
            class="text-theme-neutral-400 text-left text-xs select-none pb-3}"
            colspan="1"
            data-testid="table__th--1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="flex flex-inline align-top  undefined"
            >
              <div
                class="mr-1"
              >
                Header 2
              </div>
              <div
                class="mt-1 text-theme-color-neutral-500"
                data-testid="table__Sort"
              >
                <div
                  class="sc-AxjAm iepfoc"
                  height="10"
                  width="10"
                >
                  <svg>
                    sort.svg
                  </svg>
                </div>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr />
      </tbody>
    </table>
  </div>
</div>
`;
